### ISSUE TITLE
[Feature] [Vectorized] Some Join opt in vec exec engine

### ISSUE THREAD

**Issue by HappenLee** [apache/incubator-doris#7901]
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Description

1. shoule adjust the out join column to nullable
    `select * from t1 left join t2 on t1.a = t2.a` 
    all slots of the t2 shoule be nullable change in join node.

2. Adjust the join column of non nullaware join to non nullable
    `select * from t1 join t2 on t1.a = t2.a`  if a is a nullable column shoule change to not nullable, because of join column `=` will skip the null.

3. Useless columns do not need to be output， here need a projection to reduce the column output to speed up
the join operator.

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

**Comment by EmmyMiao87** [comment]
I will resolve this issue together ~

**Comment by EmmyMiao87** [comment]
# Join opt

## Reduce unnecessary memory copy

**Comment by Gabriel39** [comment]
还有个场景我觉得可以做类似优化，就是比如agg+having过滤，having中的列不管之后有没有用到都会带下去，事实上如果之后用不到的话，在agg这里就可以裁剪掉了

**Comment by EmmyMiao87** [comment]
This means that, in fact, the output columns of many operators may be a subset of the input columns. 
So not only the join operator can do this optimization.

**Comment by Gabriel39** [comment]
@EmmyMiao87 Exactly. Btw, just FYI, it should be highlight that which operation benefit from this. In our practice, shuffling profit most from column pruning so we only do projection (including column pruning, some computing operations) in each datastreamsender instead of each node (join, agg, etc.).

**Comment by EmmyMiao87** [comment]
>  shuffling profit most from column pruning so we only do projection (including column pruning, some computing operations) in each datastreamsender instead of each node (join, agg, etc.).

When be is implemented, it is possible to put the project logic in the sender uniformly. 
However, in the case of vectorization, if the project logic is pushed down to the hash join node, the time-consuming of **memory copying can also be reduced during node calculation**. Therefore, the project implementation of hash join is not uniformly placed in the sender.

**Comment by EmmyMiao87** [comment]
# 关于 Outer join 导致的列属性从非 null 变更到 nullable 的问题

1. 临时禁止：为了 1.0.1 版本的稳定发布，现决定当遇到这种 case 下先退回非向量化引擎执行功能。
2. 短期实现：利用变更分布式执行计划，来增加 join 的tuple 从而解决非null 变更到 nullable 的问题
3. 长期实现：通过新优化器，彻底给 join 生成一个新的output tuple，彻底解决问题。
